### PR TITLE
Type hinting for Views/Handlers

### PR DESCRIPTION
**Pull request by mgdon650**
### What is the context of this PR?
As part of an ongoing effort to enforce type hinting throughout the codebase, the PR aims to introduce type hinting throughout the routes module and propagate these hints through the classes and files referenced by the routes module.

### How to review
All existing unit and functional tests pass in the same way they do on main.
All type hinting added conforms to[ type hinting guidelines.](https://github.com/ONSdigital/eq-questionnaire-runner/blob/main/doc/python-type-hinting.md)

### Checklist
* [ ] Unit tests pass
* [ ] Functional Tests Pass
* [ ] Adheres to typing guidelines

### CODE REVIEW

File: Pipfile
Comment (katie-gardner) [inline Pipfile:75]: You can revert this change - not needed

File: Pipfile.lock
Comment (katie-gardner) [inline Pipfile.lock:4]: No need for package changes so this file can be reverted entirely

File: app/views/contexts/calculated_summary_context.py
Comment (katie-gardner) [inline app/views/contexts/calculated_summary_context.py:26]: wont let me highlight above, but as you have changed the types to no longer use `Union` and `Optional` you also need to remove those from the typing imports

File: mypy.ini
Comment (katie-gardner) [inline mypy.ini:34]: You can delete all these commented out ones now

**Review by katie-gardner — CHANGES_REQUESTED** [review]
Having a proper read through now but just picked out the main things that spring to mind on a quick skim

File: app/views/handlers/content.py
Comment (katie-gardner) [inline app/views/handlers/content.py:45]: the transform_block is passed in by the Content class which is using transformed_variants to get it - so it'll be an `ImmutableDict`

File: app/views/handlers/question.py
Comment (katie-gardner) [inline app/views/handlers/question.py:165]: action can also have a property called "params" which is a dictionary so it would be `dict[str, str | dict]` which isn't really worth writing so you can change it to just `dict | None`

File: app/views/handlers/confirm_email.py
Comment (katie-gardner) [inline app/views/handlers/confirm_email.py:156]: I've just realised this could be a `CloudTaskPublisher` or a `LogCloudTaskPublisher` see the setup.py, theres two potential options around line 370, I know we looked at this one, missed that 😁

File: app/views/handlers/block.py
Comment (katie-gardner) [inline app/views/handlers/block.py:153]: This comment doesn't quite make sense, do you mean other way round? you may need an ignore above as well

File: app/views/contexts/summary/block.py
Comment (katie-gardner) [inline app/views/contexts/summary/block.py:78]: this affects summary/question.py so the change needs to be there too

**Review by katie-gardner — CHANGES_REQUESTED** [review]
Still some errors which I think are the files you're looking at but great stuff! some comments so far 👍

File: app/views/handlers/confirm_email.py
Comment (mgdon650) [reply app/views/handlers/confirm_email.py:156]: I think it's CloudTaskPublisher as its calling Task from app.cloud_tasks.cloud_task_publishers.CloudTaskPublisher .

**Comment by berroar** [comment]
[Type hints: Add type hints to the Handlers module (2 SP)](https://trello.com/c/aRYDKMTZ/5709-type-hints-add-type-hints-to-the-handlers-module-2-sp)

**Comment by berroar** [comment]
Still 16 issues flagged by mypy that need to be resolved atm - mainly related to the changes for Location/Relationship Location

File: app/views/handlers/question.py
Comment (katie-gardner) [inline app/views/handlers/question.py:155]: you've got too many type ignores now, the `list_name` shouldn't need it

File: app/views/handlers/content.py
Comment (katie-gardner) [reply app/views/handlers/content.py:13]: still outstanding

File: app/views/handlers/question.py
Comment (katie-gardner) [reply app/views/handlers/question.py:165]: still outstanding

File: app/views/handlers/content.py
Comment (katie-gardner) [reply app/views/handlers/content.py:13]: sorry @mgdon650 - I must have been looking at the transformed_block rather than the mutable version being returned, ignore this one

File: app/questionnaire/relationship_router.py
Comment (katie-gardner) [inline app/questionnaire/relationship_router.py:33]: when would this be a `Location`?

File: app/views/handlers/submission.py
Comment (katie-gardner) [inline app/views/handlers/submission.py:59]: some missing comments for the #type ignores in this one

**Review by katie-gardner — CHANGES_REQUESTED** [review]

File: app/questionnaire/relationship_router.py
Comment (katie-gardner) [reply app/questionnaire/relationship_router.py:33]: ahh.. I see the issue, the relationship router is part of relationship question inheriting from question from blockhandler where the location is either. Hmmm, not sure if better to stay as is or type ignore

Comment (mgdon650) [reply app/questionnaire/relationship_router.py:33]: Will keep to RelationshipLocation and type ignore the error to keep consistent to what we did yesterday

File: app/views/contexts/thank_you_context.py
Comment (kylelawsonAND) [inline app/views/contexts/thank_you_context.py:86]: Missing type ignore explanation

File: app/views/handlers/list_add_question.py
Comment (kylelawsonAND) [inline app/views/handlers/list_add_question.py:39]: This is returning `dict[str, int]`?

File: app/views/handlers/list_action.py
Comment (berroar) [inline app/views/handlers/list_action.py:31]: Can be simplified to something like and not sure you need the ignore here?
```suggestion
         return bool(
            can_access_parent_location
            and self._current_location.list_name == self.parent_block["for_list"]
        )
```

File: app/views/handlers/list_action.py
Comment (berroar) [inline app/views/handlers/list_action.py:70]: I guess the comment should say why?

File: app/views/handlers/list_add_question.py
Comment (berroar) [inline app/views/handlers/list_add_question.py:39]: How come this one was changed? No comment here for explanation?

File: app/views/contexts/thank_you_context.py
Comment (mgdon650) [reply app/views/contexts/thank_you_context.py:86]: I'm not sure why this is here as there is no error when i removed it. Will change

File: app/views/handlers/list_remove_question.py
Comment (berroar) [reply app/views/handlers/list_remove_question.py:40]: Also why have we reverted from the simplified version?

File: app/views/handlers/question.py
Comment (berroar) [inline app/views/handlers/question.py:41]: Would move the comment above the call here

**Review by berroar — DISMISSED** [review]

File: app/views/handlers/block.py
Comment (mgdon650) [reply app/views/handlers/block.py:160]: Its a linting issue. Its moving comments about the place

File: app/views/handlers/list_remove_question.py
Comment (mgdon650) [reply app/views/handlers/list_remove_question.py:40]: It was throwing a "Returning Any from function declared to return bool" error

Comment (katie-gardner) [reply app/views/handlers/list_remove_question.py:40]: since mypy doesn't know what the types for the immutable dictionary are, and so doesn't know how the `__eq__ `method is implemented, but casting for the sake of type hinting is avoided, but type hinting documentation states to only type ignore where its unavoidable, could this be a middle ground?
```
list_name: str = self.parent_block["for_list"]
return list_name == self._schema.get_individual_response_list()
```
I'm wondering if we need to document some sort of agreement on this type of scenario where it's obviously a boolean, and the above is longer than the simplified version, and we _can_ avoid a type ignore, but maybe don't want to in this case?

File: app/views/handlers/list_add_question.py
Comment (mgdon650) [reply app/views/handlers/list_add_question.py:39]: It was a linting error. Have updated after Kyles comment

File: app/views/contexts/thank_you_context.py
Comment (mgdon650) [reply app/views/contexts/thank_you_context.py:86]: Think this was linting again.

File: app/views/handlers/confirmation_email.py
Comment (mgdon650) [reply app/views/handlers/confirmation_email.py:1]: Updated. Will appear in next commit

File: app/views/handlers/confirmation_email.py
Comment (mgdon650) [reply app/views/handlers/confirmation_email.py:71]: 👍

File: app/views/handlers/individual_response.py
Comment (mgdon650) [reply app/views/handlers/individual_response.py:124]: I have tried changing the code but the unit tests are failing which would mean more of an investigation into the tests/code which would be time consuming and not as critical of an issue considering individual_response isn't used.

Will leave as errors for now so that they do not get missed.

File: app/views/handlers/individual_response.py
Comment (mgdon650) [reply app/views/handlers/individual_response.py:199]: same as my response above

File: app/views/handlers/individual_response.py
Comment (mgdon650) [reply app/views/handlers/individual_response.py:726]: same as above

File: app/views/handlers/list_action.py
Comment (mgdon650) [reply app/views/handlers/list_action.py:31]: Nice! Have changed this

File: app/views/handlers/list_remove_question.py
Comment (berroar) [reply app/views/handlers/list_remove_question.py:40]: Yeah probably need to discuss this, I added this comment thinking more about not amending code for the sake of type hints, but can see where you are coming from with this hint technically being avoidable . Maybe one to raise in a tech session then and as a team we can agree whether we need to update the doc.

File: app/views/contexts/thank_you_context.py
Comment (katie-gardner) [inline app/views/contexts/thank_you_context.py:45]: If metadata is a `dict` then calling `metadata.tx_id` is going to error, might be one for @berroar to confirm as I think the change was something you both talked about? would it be better to change `metadata = get_metadata(current_user) or {}` into defaulting to another proxy or just type ignoring it?

File: tests/app/test_jinja_filters.py
Comment (katie-gardner) [inline tests/app/test_jinja_filters.py:158]: I think you've deleted this by accident

File: app/views/handlers/submission.py
Comment (katie-gardner) [inline app/views/handlers/submission.py:70]: I think it would be more correct to say that the metadata exists, in a single type ignore comment, as those are attributes that will always exist on the metadata provided the metadata is not None

File: app/views/handlers/submission.py
Comment (katie-gardner) [inline app/views/handlers/submission.py:72]: (likewise)

**Review by katie-gardner — CHANGES_REQUESTED** [review]
Just 4 minor comments from the changes

File: tests/app/test_jinja_filters.py
Comment (mgdon650) [reply tests/app/test_jinja_filters.py:158]: Hmm weird, I didn't even touch that myself....

File: app/data_models/list_store.py
Comment (MebinAbraham) [inline app/data_models/list_store.py:52]: Where do we use a non int look up?

File: app/data_models/progress_store.py
Comment (MebinAbraham) [inline app/data_models/progress_store.py:164]: Given `Location | RelationshipLocation` used like 40 times, I suggest using an type alias.

File: app/views/contexts/thank_you_context.py
Comment (MebinAbraham) [reply app/views/contexts/thank_you_context.py:45]: This method should expect a metadata proxy. It is not valid to call it without one. Therefore, we should change `metadata = get_metadata(current_user) or {}` to 
`metadata: MetadataProxy= get_metadata(current_user)  # type ignore`

File: app/views/handlers/confirmation_email.py
Comment (MebinAbraham) [inline app/views/handlers/confirmation_email.py:34]: Minor, use caps `# Type ignore:` as per doc and for consistency. Applies to all.

File: app/views/handlers/confirmation_email.py
Comment (MebinAbraham) [inline app/views/handlers/confirmation_email.py:44]: Do you need `LazyString` when you got `str`?

File: app/views/handlers/content.py
Comment (MebinAbraham) [inline app/views/handlers/content.py:33]: This never returns None?

File: app/views/handlers/individual_response.py
Comment (MebinAbraham) [inline app/views/handlers/individual_response.py:135]: This comment can be removed?

File: app/views/handlers/question.py
Comment (MebinAbraham) [inline app/views/handlers/question.py:24]: Did this need changing?. Previous one would pass with empty dict, this would error on KeyError.

File: app/views/handlers/question.py
Comment (MebinAbraham) [inline app/views/handlers/question.py:42]: This warning is valid?

File: app/views/handlers/submission.py
Comment (MebinAbraham) [reply app/views/handlers/submission.py:70]: Each individual type ignore can also be removed, if you add a line like:

```py
metadata: MetadatProxy = self._metadata  # type: ignore
```

Then using metadata.case_id tx_id etc won't raise a mypy error..

File: app/views/handlers/question.py
Comment (MebinAbraham) [inline app/views/handlers/question.py:124]: Arguments should be abstract. Unless this needs dict specific methods, use `Mapping`.

File: app/views/handlers/question.py
Comment (mgdon650) [reply app/views/handlers/question.py:24]: Mypy didn't like it but instead of changing the code I could have written the following line actually. 

```return bool(answer_action and answer_action["type"] == "RedirectToListAddBlock")```

File: app/data_models/list_store.py
Comment (mgdon650) [reply app/data_models/list_store.py:52]: In :

```app/views/handlers/relationships/unrelated_question.py:44: error: Incompatible return value type (got "Union[str, List[str]]", expected "List[str]")  [return-value]
```

File: app/views/handlers/confirmation_email.py
Comment (mgdon650) [reply app/views/handlers/confirmation_email.py:44]: When we were going through this, mypy didn't like that lazy_gettext("Confirmation email") returned a LazyString hence why we included it. Re-visiting it now, it doesn't throw and error so will remove it

**Review by MebinAbraham — APPROVED** [review]
Approving with some minor comment, will leave it to the other reviews to check the remaining ones on my behalf.

File: app/views/handlers/calculated_summary.py
Comment (mgdon650) [reply app/views/handlers/calculated_summary.py:1]: I think it's something to do with the merge. Because that was deleted ages ago? Reviewing the commit this morning as some other odd bits that have overwritten too.

File: app/views/handlers/confirmation_email.py
Comment (katie-gardner) [reply app/views/handlers/confirmation_email.py]: Have checked this is fixed 👍

File: app/views/contexts/thank_you_context.py
Comment (katie-gardner) [reply app/views/contexts/thank_you_context.py]: Checked and these have been mass updated

File: app/views/contexts/thank_you_context.py
Comment (katie-gardner) [reply app/views/contexts/thank_you_context.py:45]: @mgdon650 - this one is still outstanding I think you can actually just delete the type ignore? now you've fixed the other place it shouldn't be required

File: app/views/handlers/list_remove_question.py
Comment (katie-gardner) [reply app/views/handlers/list_remove_question.py]: Will make a note about updating this in tech session and resolve for now

File: app/forms/questionnaire_form.py
Comment (katie-gardner) [inline app/forms/questionnaire_form.py:604]: you've switched the variable name to `mutable_form_data` but forgotten to return that instead

File: app/views/handlers/individual_response.py
Comment (katie-gardner) [reply app/views/handlers/individual_response.py]: I think this one is controversial - `get_individual_response_list` is a method on `questionnaire_schema` which can indeed return None, for most schemas, however its only ever used here where it can't. I think the method ought to move, but I don't think its in the scope of the ticket, so will resolve based on it being a largely unused file and it leaves the typing in `questionnaire_schema` technically more correct... I think

**Review by katie-gardner — APPROVED** [review]

File: app/views/handlers/individual_response.py
Comment (kylelawsonAND) [reply app/views/handlers/individual_response.py:124]: I think it's worth leaving as is since the scope of this ticket would grow further - my suggestion is resolve this comment, but make a ticket to investigate it as it's own issue. Should hopefully be a small one to pick up at a later date.

File: app/views/handlers/individual_response.py
Comment (kylelawsonAND) [reply app/views/handlers/individual_response.py]: Add as part of the investigation ticket as mentioned in prev comment.

File: app/views/handlers/individual_response.py
Comment (kylelawsonAND) [reply app/views/handlers/individual_response.py]: same suggestion as prev

**Review by kylelawsonAND — APPROVED** [review]

**Pull request merged**
